### PR TITLE
[#5442] Improvement(iceberg-common): Overwrite the equals and hashCode methods to avoid frequently creating HiveClientPool instances

### PR DESCRIPTION
**Pull request by caican00**
### What changes were proposed in this pull request?

Overwrite the equals and hashCode methods to avoid frequently creating HiveClientPool instances

### Why are the changes needed?

Fix: https://github.com/apache/gravitino/issues/5442

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New UT.

### CODE REVIEW

**Comment by jerryshao** [comment]
Hi @caican00 will there be any potential UGI issue when we reuse the client instance? CC @yuqi1129 @jerqi to also take a look.

**Comment by caican00** [comment]
> Hi @caican00 will there be any potential UGI issue when we reuse the client instance? CC @yuqi1129 @jerqi to also take a look.

When Impersonation is disable, we use a super user to access hms, reusing the clients are reasonable.

When Impersonation is enabled , we obtain a HiveClientPool from the cache through a `Key` instance.  If the ugi or user_name is the same, the `Key` instance should be the same and the connection in the HiveClientPool should be taken instead of recreating a new HiveClientPool.

Now each request recreates a HiveClientPool instance, the HiveClientPool is meaningless.

Is my understanding correct? If there is any problem, please help point out. Thank you very much cc @yuqi1129

File: iceberg/iceberg-common/src/main/java/org/apache/gravitino/iceberg/common/utils/IcebergHiveCachedClientPool.java [hunk 1]
```diff
@@ -107,7 +112,13 @@ public IcebergHiveCachedClientPool(Configuration conf, Map<String, String> prope
   @VisibleForTesting
   HiveClientPool clientPool() {
     Key key = extractKey(properties.get(CatalogProperties.CLIENT_POOL_CACHE_KEYS), conf);
```
Comment (caican00) [inline iceberg/iceberg-common/src/main/java/org/apache/gravitino/iceberg/common/utils/IcebergHiveCachedClientPool.java:114]: when enabled , the `CatalogProperties.CLIENT_POOL_CACHE_KEYS` should not be null and must cantain `ugi` or `user_name`

Comment (jerryshao) [reply iceberg/iceberg-common/src/main/java/org/apache/gravitino/iceberg/common/utils/IcebergHiveCachedClientPool.java:114]: I see. Thanks for the explanation.

**Comment by yuqi1129** [comment]
> > Hi @caican00 will there be any potential UGI issue when we reuse the client instance? CC @yuqi1129 @jerqi to also take a look.
> 
> When Impersonation is disabled, we use a super user to access hms, reusing the clients is reasonable.
> 
> When Impersonation is enabled , we get a HiveClientPool instance from the clientPoolCache through a `Key` instance. If the ugi or user_name of the current thread is the same, the `Key` instance should be the same and the connection in the HiveClientPool should be taken instead of recreating a new HiveClientPool.
> 
> Now each request will create a HiveClientPool instance, the HiveClientPool is meaningless.
> 
> Is my understanding correct? If there is any problem, please help point out. Thank you very much cc @yuqi1129

Another user has reported that `Impersonation` does not work when using Spark to access Iceberg with Kerberoized Hive due to the reuse to pool,  I have not confirmed it as the environment is not accessible to me.  they claim that "Abandon the cache and just create a new pool works fine" and I'm not sure whether it's related to this PR.

**Comment by yuqi1129** [comment]
I'm okay with the current changes.

**Comment by jerryshao** [comment]
I think we can merge this first, as there's no side affect for this change, and the impersonation problem can be investigated separately.

**Review by jerryshao — APPROVED** [review]

**Comment by caican00** [comment]
> > > Hi @caican00 will there be any potential UGI issue when we reuse the client instance? CC @yuqi1129 @jerqi to also take a look.
> > 
> > 
> > When Impersonation is disabled, we use a super user to access hms, reusing the clients is reasonable.
> > When Impersonation is enabled , we get a HiveClientPool instance from the clientPoolCache through a `Key` instance. If the ugi or user_name of the current thread is the same, the `Key` instance should be the same and the connection in the HiveClientPool should be taken instead of recreating a new HiveClientPool.
> > Now each request will create a HiveClientPool instance, the HiveClientPool is meaningless.
> > Is my understanding correct? If there is any problem, please help point out. Thank you very much cc @yuqi1129
> 
> Another user has reported that `Impersonation` does not work when using Spark to access Iceberg with Kerberoized Hive due to the reuse to pool, I have not confirmed it as the environment is not accessible to me. they claim that "Abandon the cache and just create a new pool works fine" and I'm not sure whether it's related to this PR.

> > > they claim that "Abandon the cache and just create a new pool works fine" and I'm not sure whether it's related to this PR

Hi @yuqi1129 May i ask how the user abandoned the cache?

**Comment by yuqi1129** [comment]
> Hi @yuqi1129 May i ask how the user abandoned the cache?

They only tested locally and changed the method `HiveClientPool clientPool()` to create a new pool each time. Let me get a more clear information and feed back to you.

**Comment by caican00** [comment]
> > Hi @yuqi1129 May i ask how the user abandoned the cache?
> 
> They only tested locally and changed the method `HiveClientPool clientPool()` to create a new pool each time. Let me get a more clear information and feed back to you.

Thanks for you reply.

**Pull request merged**